### ISSUE TITLE
google/deepvariant:1.5.0-gpu | google/deepvariant:1.6.1-gpu docker images run as CPU-only because they are using ancient CUDA 11.3.1

### ISSUE THREAD

**Issue by s-batalov** [google/deepvariant#844]
**Have you checked the FAQ? https://github.com/google/deepvariant/blob/r1.6.1/docs/FAQ.md**:
Yes

**Describe the issue:**
google/deepvariant:1.5.0-gpu google/deepvariant:1.6.1-gpu docker images run as CPU-only because they are using ancient CUDA 11.3.1
Could maintainers build newer docker images with CUDA >=12.4 or at least >=11.8 to be able to use modern cards such as H100 and L40S (CUDA CC = 8.9 and 9.0)

**Setup**
  - Operating system: RHEL 8.10
  - DeepVariant version: 1.5.0-gpu, 1.6.1-gpu
  - Installation method (Docker, built from source, etc.): docker 
  - Type of data: (sequencing instrument, reference genome, anything special that is unlike the case studies?) Pacbio Revel fresh data. 

**Steps to reproduce:**
  - Command: docker run --gpus 1 google/deepvariant:1.5.0-gpu or docker run --gpus 1 google/deepvariant:1.6.1-gpu
  - Error trace: (if applicable)
 ...
CUDA Version 11.3.1

Container image Copyright (c) 2016-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

This container image and its contents are governed by the NVIDIA Deep Learning Container License.
By pulling and using the container, you accept the terms and conditions of this license:
https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license

A copy of this license is made available in this container at /NGC-DL-CONTAINER-LICENSE for your convenience.

2024-07-03 17:21:57.549571: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI AVX512_BF16 AVX_VNNI AMX_TILE AMX_INT8 AMX_BF16 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-07-03 17:21:57.644332: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-07-03 17:21:58.247052: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libcublas.so.12: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2024-07-03 17:21:58.247080: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
Runs all 3 steps to go from input DNA reads to output VCF/gVCF files.
(... then -- the list of all options follows)

If run on a proper BAM file with all options provided, all TF-TRT warning messages are periodically repeated as well as 
CUDA Version 11.3.1
2024-07-02 22:47:07.493311: E tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:267] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2024-07-02 22:47:12.386498: E tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:267] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
...
and processing is performed on CPUs. 

Also, these details are put in the log hudreds of times:
2024-07-03 18:27:31.862526: E tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:267] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2024-07-03 18:27:31.862557: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: 8308a7bb3067
2024-07-03 18:27:31.862563: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: 8308a7bb3067
2024-07-03 18:27:31.862607: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: 555.42.6
2024-07-03 18:27:31.862621: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 555.42.6
2024-07-03 18:27:31.862626: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:310] kernel version seems to match DSO: 555.42.6
(then repeated with every call)

**Does the quick start test work on your system?**
Please test with https://github.com/google/deepvariant/blob/r1.6/docs/deepvariant-quick-start.md.
Is there any way to reproduce the issue by using the quick start?

**Any additional context:**
L40S card is CUDA CC = 8.9
supported since CUDA >= 11.8 currently 12.4 and 12.5
https://developer.nvidia.com/cuda-11-8-0-download-archive - installed, still doesn't work with containerized 11.3.1

tensorRT is used
https://docs.nvidia.com/deeplearning/tensorrt/support-matrix/index.html 
currently 10.1 ; version 8 in CUDA 11.8
python >= 3.8

lspci | grep -i nvidia
0a:00.0 3D controller: NVIDIA Corporation AD102GL [L40S] (rev a1)
ae:00.0 3D controller: NVIDIA Corporation AD102GL [L40S] (rev a1)

nvidia-container-cli is installed (supersedes nvidia-docker - https://github.com/NVIDIA/nvidia-docker )

**Comment by kishwarshafin** [comment]
@s-batalov, I will take a look soon on how difficult it is to update the CUDA version on our end. However, it's difficult to rebuild the 1.6.1 version with an updated CUDA version, it may have to be the next release where we update the CUDA version.

**Comment by kishwarshafin** [comment]
So, as @pichuan said, this is non-actionable for now and will have to wait until next release. So I'll close this bug.

**Comment by walidabualafia** [comment]
Does this fix the mis-linked `libnvinfer_plugin.so.7` dlerror? 

Seems like both `libnvinfer_plugin.so.7` and `libnvinfer_plugin.so.8` are in `LD_LIBRARY_PATH`, but the binary `run_deepvariant` is linked against `libnvinfer_plugin.so.7`, which complains about missing `libcublas.so.12`.

Here's the error I'm getting:
```stdout
2024-10-07 09:10:29.222934: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-10-07 09:11:35.925528: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libcublas.so.12: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64:/.singularity.d/libs
2024-10-07 09:11:35.925571: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
```

Here's where I found `libnvinfer.so.7`:
```stdout
[walid@a100: 1.6.1]$ singularity exec --nv deepvariant_1.6.1-gpu.sif ls /usr/local/nvidia/lib
libnvinfer.so.7  libnvinfer.so.8  libnvinfer_plugin.so.7  libnvinfer_plugin.so.8
```

Here's my `ldd` call to see what it's linked to:
```stdout
[walid@a100: 1.6.1]$ singularity exec --nv deepvariant_1.6.1-gpu.sif ldd /usr/local/nvidia/lib/libnvinfer_plugin.so.7
	linux-vdso.so.1 (0x0000155555524000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x0000155553038000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x0000155553032000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x0000155553028000)
	libcublas.so.12 => not found
	libcublasLt.so.12 => not found
	libcudnn.so.8 => /lib/x86_64-linux-gnu/libcudnn.so.8 (0x0000155552dfc000)
	libnvinfer.so.8 => /usr/local/nvidia/lib/libnvinfer.so.8 (0x000015554454d000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x000015554436b000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x000015554421c000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x0000155544201000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000015554400f000)
	/lib64/ld-linux-x86-64.so.2 (0x0000155555526000)
```

When I grep for `libcublas` in the container:
```stdout
[walid@a100: 1.6.1]$ singularity exec --nv deepvariant_1.6.1-gpu.sif /usr/bin/grep -r -i 'libcublas' /
/.singularity.d/env/10-docker2singularity.sh:export NV_LIBCUBLAS_PACKAGE_NAME="${NV_LIBCUBLAS_PACKAGE_NAME:-"libcublas-11-3"}"
/.singularity.d/env/10-docker2singularity.sh:export NV_LIBCUBLAS_VERSION="${NV_LIBCUBLAS_VERSION:-"11.5.1.109-1"}"
/.singularity.d/env/10-docker2singularity.sh:export NV_LIBCUBLAS_PACKAGE="${NV_LIBCUBLAS_PACKAGE:-"libcublas-11-3=11.5.1.109-1"}"
/.singularity.d/env/10-docker2singularity.sh:export NV_LIBCUBLAS_DEV_VERSION="${NV_LIBCUBLAS_DEV_VERSION:-"11.5.1.109-1"}"
/.singularity.d/env/10-docker2singularity.sh:export NV_LIBCUBLAS_DEV_PACKAGE_NAME="${NV_LIBCUBLAS_DEV_PACKAGE_NAME:-"libcublas-dev-11-3"}"
/.singularity.d/env/10-docker2singularity.sh:export NV_LIBCUBLAS_DEV_PACKAGE="${NV_LIBCUBLAS_DEV_PACKAGE:-"libcublas-dev-11-3=11.5.1.109-1"}"
*** bunch more omitted output. I just wanted to show above versions ***
```

I tried to pull and bind a local install of cublas, but I did not want to make my dependencies too complicated. 

TL;DR: please test if `libcublas.so.12` is properly linked. Thank you!

Thank you all for looking into this. Please let me know if I can test or be of help. :smile:

Cheers,
Walid
